### PR TITLE
ddl: partition by range value should accept constant expression

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

In partitioned table definition, the range value maybe not an integer, it could be a constant expression.
For example, the following two cases are the same:
	PARTITION p0 VALUES LESS THAN (TO_SECONDS('2004-01-01'))
	PARTITION p0 VALUES LESS THAN (63340531200)

### What is changed and how it works?

Add a `getRangeValue` function.

### Check List 

Tests 

 - Unit test

Side effects

The `model.PartitionInfo`  may be modified.

@ciscoxll  @coocood

### CODE REVIEW

**Comment by coocood** [comment]
@tiancaiamao 
Can we keep the original expression?
We just need to know if it can be folded to a constant.

File: ddl/partition.go [hunk 1]
```diff
@@ -169,15 +171,19 @@ func checkCreatePartitionValue(pi *model.PartitionInfo) error {
...
 
-		currentRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
+		currentRangeValue, fromExpr, err := getRangeValue(ctx, tblInfo, defs[i].LessThan[0])
 		if err != nil {
-			return ErrNotAllowedTypeInPartition.GenByArgs(defs[i].LessThan[0])
+			return errors.Trace(err)
+		}
+		if fromExpr {
+			// Constant fold the expression.
+			defs[i].LessThan[0] = fmt.Sprintf("%d", currentRangeValue)
```
Comment (winkyao) [inline ddl/partition.go]: Why not keep the original expression?

Comment (winkyao) [reply ddl/partition.go]: Got.

**Comment by coocood** [comment]
LGTM

**Pull request merged**